tensorflow/core/kernels/save_restore_tensor.cc

```diff
@@ -250,13 +250,23 @@ Status RestoreTensorsV2(OpKernelContext* context, const Tensor& prefix,
   // TODO(zongheng): consider measuring speed and issuing concurrent lookups
   // within a fixed memory budget.
   TensorShape restored_full_shape;
+  DataType original_dtype;
   Tensor* restored_tensor = nullptr;
   for (auto i : sorted_name_idx) {
     const string& tensor_name = tensor_names_flat(i);
     const string& shape_and_slice = shape_and_slices_flat(i);
 
     TF_RETURN_IF_ERROR(
-        reader.LookupTensorShape(tensor_name, &restored_full_shape));
+        reader.LookupDtypeAndShape(tensor_name,
+                                   &original_dtype, &restored_full_shape));
+
+    // Precheck dtype.
+    if (dtypes[i] != original_dtype) {
+      return errors::InvalidArgument(
```

Comment:
It might be more user-friendly to collect all variables that have a mismatch and then report them all.